### DISCUSSION TITLE
New to Magma, Deploy full magma architecure on AWS

### DISCUSSION THREAD

**Discussion by vik2595** [magma/magma#6201] [Q&A]
Hi Everyone, 

I'm new to magma,

Is there anyone who already deployed full magma architecture on AWS? 

I'm following `magma/experimental/cloudstrapper` to deploy this on AWS. but not successful so far.

Any suggestions would be appreciate.

**Comment by karthiksubraveti** [comment]
@vik2595 If you post the errors I can see how I can help here. 
Meanwhile, I have added orcl https://github.com/magma/magma/tree/master/orc8r/cloud/deploy/orc8r_deployer/docker to deploy orc8r. Plan to make this as a single interface to deploy orc8r across the board. Currently it works well with aws/terraform. 

With orcl, you have to run through following steps

> orcl configure 
Lets you configure all mandatory config attributes

> orcl certs add --self-signed
Will add certs into a specific secrets directory

> orcl install
Will walk through install prechecks followed by actual installation

Let me know if you see any issues here.

**Reply by vik2595** [reply]
My work so far
Infrastructure:
- EC2 instance with t3.medium
- OS: Ubuntu 20.04
- EBS: 30 GB
- region: ca-central-1

Installation: 
- installed prerequisite modules (mentioned on [Magma prerequisites](https://docs.magmacore.org/docs/basics/prerequisites) )
- updated roles/vars/secrets.yaml
- created BootKey and HostKey via `aws-prerequisites.yaml`

I'm getting following error while running `ansible-playbook devops-provision.yaml -e "dirLocalInventory=${WORK_DIR}"` 

```
root@ip-172-31-32-28:/home/ubuntu/magma/experimental/cloudstrapper/playbooks# ansible-playbook devops-provision.yaml -e "dirLocalInventory=${WORK_DIR}"
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not
match 'all'

PLAY [localhost] ***************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************
ok: [localhost]

TASK [devops-infra : query all security groups] ********************************************************************
ok: [localhost]

TASK [devops-infra : assign security group id to variable] *********************************************************
fatal: [localhost]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: list object has no element 0\n\nThe error appears to be in '/home/ubuntu/magma/experimental/cloudstrapper/playbooks/roles/devops-infra/tasks/main.yaml': line 8, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: assign security group id to variable\n  ^ here\n"}

PLAY RECAP *********************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0 
```

Any thoughts on this ? @karthiksubraveti @karthiksubraveti

**Reply by vik2595** [reply]
Also please advise, what's ami ID of cloudstrapper AMI ?

**Reply by arunuke** [reply]
@vik2595 The error you are running into is because there is no matching security group as listed in [defaults.yaml](https://github.com/arunuke/magma/blob/master/experimental/cloudstrapper/playbooks/roles/vars/defaults.yaml). [aws-prerequisites.yaml](url) should have created it. Can you share the command you ran or the output from that? Alternatively, you can check CloudFormation service to see if you have the "Essentials" stack running. If the stack is successfully running, you should have a security group created as defined in defaults which is used by the Cloudstrapper devops process.

In any case, if you have already installed all the prerequisites, you can skip the optional CI/CD steps described in sec 1.1 and move on to building the artifacts (Sec 3). Once fully built, you can choose to go to Sec 4 and deploy or check out the orcl install that @karthiksubraveti mentioned earlier. 

Keep us posted on your progress and we would be happy to assist.

**Reply by arunuke** [reply]
> Also please advise, what's ami ID of cloudstrapper AMI ?

@vik2595 If you created a Cloudstrapper AMI using the devops process outlined in Sec 1.1, you would obtain a Cloudstrapper AMI id. However, the system does not require the AMI id and self-detects it automatically based on a tag used. You can examine the [code](https://github.com/arunuke/magma/blob/master/experimental/cloudstrapper/playbooks/roles/cloudstrapper-infra/tasks/main.yaml) that does this. When Magma on AWS Marketplace goes GA, you would also see one in the public Marketplace. 

However, based on your description above, you should be able to skip the process since you have already installed all the necessary prerequisites.

**Comment by karthiksubraveti** [comment]
Btw this is our official install guide https://docs.magmacore.org/docs/orc8r/deploy_install. This is for installing orc8r on aws using terraform.

**Comment by arunuke** [comment]
I wanted to +1 Karthik's response. @vik2595 If you can specify the issues you are running into, Karthik and I can assist. Is there a specific step that is failing for you in either the orchestrator or AGW deployment via the Cloudstrapper?

**Comment by vik2595** [comment]
Thank you for install guide @karthiksubraveti 

about `magma/experimental/cloudstrapper`

what should be the BootKey and HostKey specification, 
can we use any SSH key or is it a specific key generated by cloudstrapper?

**Reply by arunuke** [reply]
@vik2595 - For the Cloudstrapper, BootKey is the one that is used for the Cloudstrapper image (if you are generating one on your own) and HostKey is the one that is used for all EC2 instances spawned by Cloudstrapper (for devops or AGW). It can be a pre-existing key, in which case you would change the parameter file, or something you can create anew through the aws-essentials role via the aws prerequisites playbook.

**Reply by vik2595** [reply]
I've successfully completed till following command. 

`ansible-playbook build-provision.yaml --tags devopsOrc8r,inventory`

What should be the content of common_instance_aws_ec2.yaml ? is there any sample available ?

**Reply by arunuke** [reply]
@vik2595 the .yaml file is an ec2 plugin that is used by ansible to work with AWS instances with one example being the ability to filter instances based on tags (instead of, say, trying to find the IP address). This file will be automatically generated when you use the 'inventory' tag for the provision playbooks by retrieving information from your secrets.yaml and cluster.yaml files.

---

plugin: aws_ec2
aws_access_key: <your key goes here>
aws_secret_key: <secret key goes here>

regions:
  - <your region goes here>

keyed_groups:
  - key: <the key you are indexing on, ex: tags>
    prefix: <the prefix you want to associate with the filter. I have used the word 'tag'>

**Comment by vik2595** [comment]
I'd installed AGW on ubuntu based on AWS. 

but getting following errors while `service magma@* status` , 

All other services are fine but, magmad service giving following errors.


```
● magma@magmad.service - Magma magmad service
     Loaded: loaded (/etc/systemd/system/magma@magmad.service; enabled; vendor preset: enabled)
     Active: active (running) since Sat 2021-05-01 03:50:25 UTC; 39s ago
   Main PID: 34608 (python3)
      Tasks: 8 (limit: 4638)
     Memory: 70.4M (limit: 300.0M)
     CGroup: /system.slice/system-magma.slice/magma@magmad.service
             └─34608 python3 -m magma.magmad.main

May 01 03:50:28 magma magmad[34608]: ERROR:root:GetServiceInfo Error for ctraced! [StatusCode.UNAVAILABLE] failed to connect to all addresses
May 01 03:50:28 magma magmad[34608]: ERROR:root:GetOperationalStates Error for ctraced! [StatusCode.UNAVAILABLE] failed to connect to all addresses
May 01 03:50:28 magma magmad[34608]: ERROR:root:GetServiceInfo Error for health! [StatusCode.UNAVAILABLE] failed to connect to all addresses
May 01 03:50:28 magma magmad[34608]: ERROR:root:GetOperationalStates Error for health! [StatusCode.UNAVAILABLE] failed to connect to all addresses
May 01 03:50:29 magma magmad[34608]: ERROR:root:Checkin Error! Failed to report states. [StatusCode.UNAVAILABLE] failed to connect to all addresses
May 01 03:50:29 magma magmad[34608]: INFO:root:StateReporting (Checkin) failure threshold met, remediating...
May 01 03:50:29 magma magmad[34608]: ERROR:root:Error making TCP connection: Unknown error -2, [Errno -2] Name or service not known
May 01 03:50:29 magma magmad[34608]: ERROR:root:StateReporting failure likely not due to invalid cert
May 01 03:50:57 magma magmad[34608]: ERROR:root:GetChallenge error! [StatusCode.UNAVAILABLE] DNS resolution failed for service: bootstrapper-controller.magma.test:7444
May 01 03:50:57 magma magmad[34608]: INFO:root:Retrying bootstrap in 30 seconds
```

Any hint on this ?

**Reply by emakeev** [reply]
magmad is trying to connect to 'bootstrapper-controller.magma.test', you either need to make sure that you have host/DNS entry for it or change it to your valid orchestrator address in your gateway's /var/opt/magma/configs/control_proxy.yml

https://docs.magmacore.org/docs/lte/deploy_config_agw#access-gateway-configuration

**Reply by arunuke** [reply]
Thanks @emakeev.

@vik2595 Can you please confirm if you have a fully functional Orc8r setup already running?
